ledger/participant-integration-api/src/main/scala/platform/store/backend/UpdateToDBDTOV1.scala

```diff
@@ -139,6 +139,11 @@ object UpdateToDBDTOV1 {
         )
 
       case u: Update.TransactionAccepted =>
+        // TODO add support for Daml Exceptions / Rollback nodes
+        //   Covering this functionality with unit test is important, since at the time of writing kvutils ledgers purge RollBack nodes already on WriteService, so conformance testing is impossible
```

Comment:
> kvutils ledgers purge RollBack nodes already on WriteService

Out of interest, how does the purging work? Reads below rollback nodes can lead to divulgence (if I understand @moritzkiefer-da correctly [here](https://github.com/digital-asset/daml/pull/9486#discussion_r620353948)), you can't just remove rollback nodes from a transaction.

If the kvutils purging operation removes create/exercise events below rollback nodes but preserves divulgence (or more generally, behaves correctly), can we apply the same operation to any incoming transaction? 

> conformance testing is impossible

I assume this means "conformance testing is impossible on kvutils ledgers in their current form", because below you note that sandbox-classic on append-only will be able to use conformance tests added in https://github.com/digital-asset/daml/pull/9400?

Reply:
This is not actually specific to rollback nodes. KVUtils also drops fetch & lookup by key nodes which have the same issue (meaning they also have the same testing issue). The logic for this is in https://github.com/digital-asset/daml/blob/c85da6def052fa3d5a817789b66b3ebe3815c9e1/ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/transaction/TransactionCommitter.scala#L383.

The reason why this doesn’t cause issues around divulgence is that KVUtils calculates blinding info before trimming nodes and then passes it along with the transaction so while the transaction has lost the information it is still there and passed into the indexer somehow (not familiar with the details there myself).

As for sandbox classic, you are completely correct. That will result in this code path being exercised.

Reply:
> KVUtils calculates blinding info before trimming nodes and then passes it along with the transaction

I was originally thinking whether we can extract this transformation (something like `TransactionWithAllKindsOfNodes => (TransactionWithWriteNodesOnly, BlindingInfo)`?) from `kvutils` into `com.daml.lf.transaction`, so that this transformation can be applied to all transactions in the indexer without having to re-implement a rollback-node-aware traversal using `foldInExecutionOrder`.

Thanks for the explanation, I was mainly interested in whether this can be done, I did not have a look at whether it makes sense.

Reply:
Definitely in favor of trying to share this. We already have a kvutils utilties library in //daml-lf that erases rollbacks https://github.com/digital-asset/daml/blob/main/daml-lf/kv-transaction-support/src/main/scala/com/daml/lf/kv/TransactionNormalizer.scala. Having something more general that pairs this with blindinginfo seems sensible to me.